lib/charms/operator_libs_linux/v0/juju_systemd_notices.py

```diff
@@ -0,0 +1,476 @@
+    task to run in the current running loop.
+
+    Args:
+        loop: Asynchronous event loop from main thread.
+    """
+    loop.call_soon(_async_load_services)
+
+
+async def _juju_systemd_notices_daemon() -> None:
+    """Start Juju systemd notices daemon.
+
+    This start call will set up the notices service to listen for events.
+    It connects to the system message bus and registers for signals/events on the
+    org.freedesktop.systemd1.Unit object looking for any PropertyChanged events.
+    This method additionally sets up signal handlers for various signals to either
+    terminate the process or reload the configuration from the hooks directory.
+    """
+    stop_event = asyncio.Event()
+    loop = asyncio.get_event_loop()
+    loop.add_signal_handler(signal.SIGINT, stop_event.set)
+    loop.add_signal_handler(signal.SIGTERM, stop_event.set)
+    # The event loop must be early bound to the lambda, otherwise the event loop
+    # will not exist within the lambda when the SIGHUP signal is received
+    # by the running notices daemon.
+    loop.add_signal_handler(  # pragma: no branch
+        signal.SIGHUP, lambda loop=loop: _load_services(loop)
+    )
+
+    sysbus = await MessageBus(bus_type=BusType.SYSTEM).connect()
+    await _async_load_services()
+
+    reply = await sysbus.call(
+        Message(
+            destination="org.freedesktop.DBus",
+            path="/org/freedesktop/DBus",
+            interface="org.freedesktop.DBus",
+            member="AddMatch",
+            signature="s",
+            body=[
+                "path_namespace='/org/freedesktop/systemd1/unit',type='signal',"
+                "interface='org.freedesktop.DBus.Properties'"
+            ],
+            serial=sysbus.next_serial(),
+        )
+    )
+    assert reply.message_type == MessageType.METHOD_RETURN
+    sysbus.add_message_handler(_systemd_unit_changed)
+    await stop_event.wait()
+
+
+def _main():
+    """Invoke the Juju systemd notices daemon.
+
+    Note:
+        This method is used to start the Juju systemd notices daemon when
+        juju_systemd_notices.py is executed as a script, not imported as a module.
+    """
+    parser = argparse.ArgumentParser()
+    parser.add_argument("-d", "--debug", action="store_true")
+    parser.add_argument("unit", type=str)
+    try:
+        args = parser.parse_args()
+        # Intentionally set as global.
+        global _juju_unit
+        _juju_unit = args.unit
+
+        console_handler = logging.StreamHandler()
+        if args.debug:
+            _logger.setLevel(logging.DEBUG)
+            console_handler.setLevel(logging.DEBUG)
+        else:
+            _logger.setLevel(logging.INFO)
+            console_handler.setLevel(logging.DEBUG)
+
+        _logger.addHandler(console_handler)
+        _logger.info("Starting juju systemd notices service")
+        asyncio.run(_juju_systemd_notices_daemon())
+    except argparse.ArgumentError:
```

Comment:
Isn't this the default for `ArgumentParser`? To print a usage message and exit? In my experience you don't usually need this try/except in scripts because of the default.

Reply:
The big thing is more or less that I want to emit an exit code 2. That way if systemd reports `juju-systemd-notices` as failed, we will know that the daemon failed due to misuse of arguments rather than that it encountered a fatal error when running. After running a faulty test without the try-except block, it does look like the exit code is 2 when you do not specify the unit name, so I will make this change.